File: pipeline/models/races.py
"""Pydantic model for races.json."""

from __future__ import annotations

from pydantic import field_validator

from .common import LenientModel, SkillBonusEntry, Source, StrictModel


class CharBonusChoice(StrictModel):
    """Characteristic bonus with choice options."""

    options: list[str]
    description: str


class RacialPowerOption(StrictModel):
    """A structured option within a racial power (e.g. Dryad choices)."""

    name: str
    description: str


class RacialPower(LenientModel):
    """A racial power with name and description.

    Most races have no options. Some (e.g. Dryad) have structured
    option objects with name + description.
    """

    name: str
    description: str
    options: list[str | RacialPowerOption] | None = None


class Race(LenientModel):
    """A playable race.

    Some races have an extra 'notes' field (e.g. Halfling).
    """

    id: str
    name: str
    size: int
    languages: list[str]
    charBonus: CharBonusChoice
    skillBonus: list[SkillBonusEntry]
    power: RacialPower
    source: Source
    notes: str | None = None

    @field_validator("size")
    @classmethod
    def size_in_range(cls, v: int) -> int:
        if not 1 <= v <= 7:
            msg = f"Size must be 1-7, got {v}"
            raise ValueError(msg)
        return v


class RacesFile(StrictModel):
    """Top-level structure of races.json."""

    races: list[Race]

File: src/lib/dtd/schemas/races.ts
/**
 * Zod schemas for race data.
 *
 * Mirrors pipeline/models/races.py.
 */

import { z } from "zod";
import { SkillBonusEntry, Source } from "./common.js";

// ---------------------------------------------------------------------------
// Sub-schemas
// ---------------------------------------------------------------------------

export const CharBonusChoice = z
	.object({
		options: z.array(z.string()),
		description: z.string(),
	})
	.strict();
export type CharBonusChoice = z.infer<typeof CharBonusChoice>;

export const RacialPowerOption = z
	.object({
		name: z.string(),
		description: z.string(),
	})
	.strict();
export type RacialPowerOption = z.infer<typeof RacialPowerOption>;

export const RacialPower = z
	.object({
		name: z.string(),
		description: z.string(),
		options: z.array(z.union([z.string(), RacialPowerOption])).nullish(),
	})
	.passthrough();
export type RacialPower = z.infer<typeof RacialPower>;

// ---------------------------------------------------------------------------
// Top-level schemas
// ---------------------------------------------------------------------------

export const Race = z
	.object({
		id: z.string(),
		name: z.string(),
		size: z.number().int().min(1).max(7),
		languages: z.array(z.string()),
		charBonus: CharBonusChoice,
		skillBonus: z.array(SkillBonusEntry),
		power: RacialPower,
		source: Source,
		notes: z.string().nullish(),
	})
	.passthrough();
export type Race = z.infer<typeof Race>;

export const RacesFile = z
	.object({
		races: z.array(Race),
	})
	.strict();
export type RacesFile = z.infer<typeof RacesFile>;
